File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [[0]*n for _ in range(m)]

        for r in range(m):
            for c in range(n):
                if r==0 and c==0:
                    dp[r][c]=grid[r][c]
                elif r==0:
                    dp[r][c]=grid[r][c]+dp[r][c-1]
                elif c==0:
                    dp[r][c]=grid[r][c]+dp[r-1][c]
                else:
                    dp[r][c]=grid[r][c]+min(dp[r-1][c],dp[r][c-1])

        return dp[m-1][n-1]

File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n,0));
        for (int r = 0; r < m; r++) {
            for (int c =0; c<n; c++){
                if(r==0 && r==c){
                    dp[r][c] = grid[r][c];
                } else if(c==0){
                    dp[r][c]=grid[r][c]+dp[r-1][c];
                } else if(r==0){
                    dp[r][c]=grid[r][c]+dp[r][c-1];
                } else {
                    dp[r][c]=grid[r][c]+min(dp[r-1][c],dp[r][c-1]);
                }
            }
        }
    return dp[m-1][n-1];
    }
};